src/radical/pilot/utils/serializer.py

```diff
@@ -0,0 +1,159 @@
+import tempfile
+
+import radical.utils as ru
+
+
+class Serializer(object):
+
+    """
+    The serializer should be able to (de)serialize information that we want
+    to send over the wire from the client side to the agent side via
+    1- ZMQ
+    2- MongoDB
+    """
+
+
+    def __init__(self):
+        """ Instantiate the class
+        """
+        self._log           = ru.Logger(name='serializer', level='DEBUG')
+        self._obj_dir       = tempfile.gettempdir()
+        self._obj_file_name = 'rp_obj.pkl'
+        self._obj_file_path = os.path.join(self._obj_dir, self._obj_file_name)
+
+
+    def serialize_obj(self, obj):
+        """
+        serialize object
+        """
+        result = None
+        exception = None
+
+        if callable(obj):
+            try:
+                result = dill.dumps(obj)
+            except Exception as e:
+                exception = e
+                self._log.exception("failed to serialize object: %s", e)
+
+            if not result:
+                # see issue: https://github.com/uqfoundation/dill/issues/128
+                try:
+                    result = dill.dumps(obj, byref = True)
+                except Exception as e:
+                    exception = e
+                    self._log.exception("failed to serialize object (by ref): %s", e)
+
+        else:
+            try:
+                result = dill.dumps(obj, recurse = True)
+            except Exception as e:
+                exception = e
+                self._log.exception("failed to serialize object: %s", e)
+
+            if not result:
+                # see issue: https://github.com/uqfoundation/dill/issues/128
+                try:
+                    result = dill.dumps(obj, byref = True)
+                except Exception as e:
+                    exception = e
+                    self._log.exception("failed to serialize object (by ref): %s", e)
+
+        if result is None:
+            raise Exception("object %s is not serializable", exception)
+
+        return  result
+
+
+    def serialize_file(self, obj):
+        """
+        serialize object to file
+        """
+        result = None
+        exception = None
+
+        if callable(obj):
+            try:
+                with open(self._obj_file_path, 'wb') as f:
+                    dill.dump(obj, f)
+                    result = self._obj_file_path
+
+            except Exception as e:
+                exception = e
+                self._log.exception("failed to serialize object to file: %s", e)
+
+        else:
+            try:
+                with open(self._obj_file_path, 'wb') as f:
+                    dill.dump(obj, f, recurse = True)
+                    result = self._obj_file_path
+            except Exception as e:
+                exception = e
+                self._log.exception("failed to serialize object to file: %s", e)
+
+        if result is None:
+            raise Exception("object is not serializable: %s", exception)
```

Comment:
This should be:
```py
raise Exception("object is not serializable") from exception
```
to preserve the original exception stack for debugging.

Reply:
fixed